src/mlpack/core/tree/binary_space_tree/rp_tree_mean_split_impl.hpp

```diff
@@ -0,0 +1,191 @@
+{
+  const size_t maxNumSamples = 100;
+  const size_t numSamples = std::min(maxNumSamples, count);
+  arma::uvec samples;
+
+  // Get no more than numSamples distinct samples.
+  GetDistinctSamples(samples, begin, count, numSamples);
+
+  // Find the average distance between points.
+  ElemType averageDistanceSq = GetAveragePointDistance(data, samples);
+
+  const ElemType threshold = 10;
+
+  if (bound.Diameter() * bound.Diameter() <= threshold * averageDistanceSq)
+  {
+    // We will perform the median split.
+    splitInfo.meanSplit = false;
+
+    splitInfo.direction.zeros(data.n_rows);
+
+    // Get a random normal vector.
+    GetRandomDirection(splitInfo.direction);
+
+    // Get the median value of the scalar products of the normal and the
+    // sampled points. The node will be split according to this value.
+    if (!GetDotMedian(data, samples, splitInfo.direction, splitInfo.splitVal))
+      return false;
+  }
+  else
+  {
+    // We will perform the mean split.
+    splitInfo.meanSplit = true;
+
+    // Get the median of the distances between the mean point and the sampled
+    // points. The node will be split according to this value.
+    if (!GetMeanMedian(data, samples, splitInfo.mean, splitInfo.splitVal))
+      return false;
+  }
+
+  return true;
+}
+
+template<typename BoundType, typename MatType>
+void RPTreeMeanSplit<BoundType, MatType>::GetDistinctSamples(
+    arma::uvec& distinctSamples,
+    const size_t begin,
+    const size_t count,
+    const size_t numSamples)
+{
+  arma::Col<size_t> samples;
+
+  samples.zeros(count);
+
+  for (size_t i = 0; i < numSamples; i++)
+    samples [ (size_t) math::RandInt(count) ]++;
+
+  distinctSamples = arma::find(samples > 0);
+
+  distinctSamples += begin;
+}
+
+template<typename BoundType, typename MatType>
+typename MatType::elem_type RPTreeMeanSplit<BoundType, MatType>::
+GetAveragePointDistance(
+    MatType& data,
+    const arma::uvec& samples)
+{
+  ElemType dist = 0;
+
+  for (size_t i = 0; i < samples.n_elem; i++)
+    for (size_t j = i + 1; j < samples.n_elem; j++)
+      dist += metric::SquaredEuclideanDistance::Evaluate(data.col(samples[i]),
+          data.col(samples[j]));
+
+  dist /= (samples.n_elem * (samples.n_elem - 1) / 2);
+
+  return dist;
+}
+
+template<typename BoundType, typename MatType>
+void RPTreeMeanSplit<BoundType, MatType>::GetRandomDirection(
+    arma::Col<ElemType>& direction)
+{
+  arma::Col<ElemType> origin;
+
+  origin.zeros(direction.n_rows);
+
+  for (size_t k = 0; k < direction.n_rows; k++)
+    direction[k] = math::Random(-1.0, 1.0);
+
+  ElemType length = metric::EuclideanDistance::Evaluate(origin, direction);
+
+  if (length > 0)
+    direction /= length;
+  else
+  {
+    // If the vector is equal to 0, choose an arbitrary dimension.
+    size_t k = math::RandInt(direction.n_rows);
+
+    direction[k] = 1.0;
+
+    length = metric::EuclideanDistance::Evaluate(origin, direction);
+
+    direction[k] /= length;
+  }
+}
```

Comment:
I think this also can be replaced with the shorter function idea I sketched in another comment.  Another thought is, since you are using it twice, maybe mark the function static in `RPTreeMaxSplit` and then call it from `RPTreeMeanSplit` instead of implementing it again here.  I am not sure this is a great function to add to the core mlpack code, because of its odd behavior when a vector of length 0 is selected.  Is that behavior the best thing to do here, or maybe is it better to simply draw a random vector again?